Swap promises for await

michaelprichardson opened the pull request:
Replaced the promise arrays with await/async. This helped to remove the `sleep()` in the tests

michaelprichardson committed:
refactor(deleteReferences): Swap to await and add types

michaelprichardson committed:
refactor(replicateAttributes): Swap to await

michaelprichardson committed:
refactor(maintainCount): Swap to await

michaelprichardson committed:
test: Swap to await

michaelprichardson committed:
test(unit): Remove sleep

nbransby commented on src/rules/deleteReferences.ts:
```diff
@@ -107,19 +114,15 @@ export function integrifyDeleteReferences(
-          })
-        );
-      });
-      return Promise.all(promises);
```
Might still be worth using Promise.all to get these queries to run in parallel as now they will be sequential.

nbransby commented on src/rules/deleteReferences.ts:
```diff
@@ -107,19 +114,15 @@ export function integrifyDeleteReferences(
+              target.isCollectionGroup ? ' (group)' : ''
+            }, id [${doc.id}]`
+          );
+          await doc.ref.delete();
```
it deletes each record one by one? Should really use a WriteBatch here

nbransby commented on src/rules/replicateAttributes.ts:
```diff
@@ -121,26 +121,19 @@ export function integrifyReplicateAttributes(
+          );
+          await doc.ref.update(update);
+        }
+      }
```
same points apply here

nbransby commented on src/common.ts:
```diff
@@ -10,6 +17,11 @@ export interface Config {
+export type PreHookFunction = (
+  change: DocumentSnapshot | Change<QueryDocumentSnapshot>,
+  context: functions.EventContext
+) => Promise<void>;
```
Consider making the return type ` Promise<void> | void` that way clients dont have to use the async keyword if they usign await

michaelprichardson committed:
fix(replicateAttributes): Use batch update

michaelprichardson committed:
fix(deleteReferences): Use batch delete

michaelprichardson committed:
fix(unit): Change check to await

michaelprichardson committed:
chore(package): Updated to version 4.0.1

nbransby reviewed (commented):
looks good. you might want to copy our WriteBatch class in and use that instead as this will blow up if there is more than 500 records in the collection

michaelprichardson committed:
feat(WriteBatch): Add write batch class

michaelprichardson committed:
fix(package): Add missing dependencies

michaelprichardson commented:
Added the WriteBatch class and replaced the standard batch writes in delete and replicate. There was an issue with the import of `object-sizeof`, had to replace `sizeof` with `require('object-sizeof')`

michaelprichardson committed:
feat(deleteReferences): Change dynamic variable of snapshot to

michaelprichardson committed:
test(deleteReferences): Change dynamic variable of snapshot to

michaelprichardson committed:
chore(readme): Update readme

nbransby commented on test/unit.test.js:
```diff
@@ -409,6 +409,7 @@ async function testDeleteReferences(sut, t, name) {
     t.is(state.context.params.masterId, masterId);
   }
 
+  console.log('Here 1');
```
guessing this can be removed now

nbransby commented on test/functions/index.js:
```diff
@@ -33,7 +33,7 @@ module.exports.replicateMasterToDetail = integrify({
   ],
   hooks: {
-    pre: (change, context) => {
+    pre: async (change, context) => {
```
can async be removed from here now or is setState async?

michaelprichardson committed:
fix(unit): Cleaned up test files
